### PR TITLE
Improve Protocol.UndefinedError message...

### PR DESCRIPTION
**Pull request by treble37**
when passing the module name of a struct

Addresses https://github.com/elixir-lang/elixir/issues/8532

EDIT: Just took a stab at this issue based on the comments I read, I was hoping to find a better way to infer the type of value, but I used guard clauses, open to suggestions of how to improve this if any....

### CODE REVIEW

File: lib/elixir/lib/exception.ex [hunk 1]
```diff
@@ -1150,9 +1150,21 @@ defmodule Protocol.UndefinedError do
...
   @impl true
   def message(%{protocol: protocol, value: value, description: description}) do
-    "protocol #{inspect(protocol)} not implemented for #{inspect(value)}" <>
-      maybe_description(description) <> maybe_available(protocol)
-  end
+    "protocol #{inspect(protocol)} not implemented for #{inspect(value)} of type " <>
+      value_type(value) <> maybe_description(description) <> maybe_available(protocol)
+  end
+
+  defp value_type(value) when is_bitstring(value), do: "BitString"
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex]: @treble37 this looks excellent, we are just missing the `__struct__` type. Something like:

    defp value_type(%{__struct__: struct}) when is_atom(struct), do: inspect(struct)

Can you please add this clause and add a test accordingly if we don't have any? Thank you.

Comment (treble37) [reply lib/elixir/lib/exception.ex]: Will do, thank you for the feedback!

Comment (treble37) [reply lib/elixir/lib/exception.ex]: Updated @josevalim 😃

File: lib/elixir/lib/exception.ex [hunk 2]
```diff
@@ -1150,10 +1150,26 @@ defmodule Protocol.UndefinedError do
 
   @impl true
   def message(%{protocol: protocol, value: value, description: description}) do
-    "protocol #{inspect(protocol)} not implemented for #{inspect(value)}" <>
-      maybe_description(description) <> maybe_available(protocol)
+    "protocol #{inspect(protocol)} not implemented for #{inspect(value)} of type " <>
+      value_type(value) <> maybe_description(description) <> maybe_available(protocol)
   end
 
+  defp value_type(value) when is_bitstring(value), do: "BitString"
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/elixir/lib/exception.ex]: We don't have a separate type for binary/strings on protocol dispatch, they are all BitString.

[the remark this answers is not in the document]
Comment (eksperimental) [reply lib/elixir/lib/exception.ex]: Good to know. I wasn't sure about it

File: lib/elixir/lib/exception.ex [hunk 3]
```diff
@@ -1150,10 +1150,26 @@ defmodule Protocol.UndefinedError do
...
+  defp value_type(value) when is_atom(value), do: "Atom"
+  defp value_type(value) when is_function(value), do: "Function"
+  defp value_type(value) when is_list(value), do: "List"
+  defp value_type(value) when is_map(value), do: "Map"
+  defp value_type(value) when is_pid(value), do: "PID"
+  defp value_type(value) when is_float(value), do: "Float"
+  defp value_type(value) when is_integer(value), do: "Integer"
+  defp value_type(value) when is_port(value), do: "Port"
+  defp value_type(value) when is_tuple(value), do: "Tuple"
+  defp value_type(value) when is_reference(value), do: "Reference"
```
[the remark this answers is not in the document]
Comment (treble37) [reply lib/elixir/lib/exception.ex]: Sure, updated @eksperimental, I had to keep the "struct" match before the "map" match for the tests to pass.

**Comment by michalmuskala** [comment]
Thank you! :heart:

**Pull request merged**